### PR TITLE
Upstream merge

### PR DESCRIPTION
**Pull request by daringer**
merge notes/todos:

* `dispatch-fido` is not needed anymore, thus remove (included in `fido-auth` with "dispatch" in features)
* `fido-authenticator` has been taken from `upstream/main` for now -> @todo ?
  * or does version 0.1 already contain all things we need ?
  * so do we merge `fido-authenticator` upstream, or stay on version = 0.1 
* `ctaphid.set_version(usbd_ctaphid::Version {` 
  * don't we want that?
  * in `initilizer.rs` 
  * added for now [here](https://github.com/Nitrokey/nitrokey-3-firmware/blob/37338c294bbdc117c2b798694edaab900fc155c8/runners/lpc55/src/initializer.rs#L537)
* inside `initializer.rs` `get_serial_number()` 
  * afaik this should not be set for fido2-devices
  * did not take the upstream implementation, just returning "00....00"
* various RGB related changes and adaptations inside `board/trussed.rs` 
  * to be discussed if we really want a math-lib as dep to calc a LED-amplitude
  * anyways RGB should be heavily tested
  * added `Intensities::scale_by()` for easy scaling (currently misused by scaling with the amplitude...)
* `src/types.rs` `UsbClasses` with or without `serial` ?
  * maybe `#cfg[]` ? based on feature?
  * `serial` pulled in for now...

### CODE REVIEW

File: runners/lpc55/src/initializer.rs [hunk 1]
```diff
@@ -72,6 +72,18 @@ pub struct Initializer {
...
+    use core::fmt::Write;
+    unsafe {
+        let uuid = crate::hal::uuid();
+        SERIAL_NUMBER.write_fmt(format_args!("{}", hexstr!(&uuid))).unwrap();
+        &SERIAL_NUMBER
+    }*/
+    "00000000-0000-0000-00000000"
+}
+
+
```
Comment (szszszsz) [inline runners/lpc55/src/initializer.rs:86]: IMO showing serial number should be configurable, with the default being the SN exposed to allow managing devices in the enterprise setting. Let's discuss this.

**Comment by szszszsz** [comment]
> dispatch-fido is not needed anymore, thus remove (included in fido-auth with "dispatch" in features)
    fido-authenticator has been taken from upstream/main for now -> @todo ?
        or does version 0.1 already contain all things we need ?
        so do we merge fido-authenticator upstream, or stay on version = 0.1

AFAIK `fido-authenticator` is updated with our changes, but potentially modified, and with that breaking user data compatibility. Let's use the released version. To do: check data retention for the FIDO2 key handles; verify changes against our last used.

>    ctaphid.set_version(usbd_ctaphid::Version {
        don't we want that?
        in initilizer.rs
        added for now [here](https://github.com/Nitrokey/nitrokey-3-firmware/blob/37338c294bbdc117c2b798694edaab900fc155c8/runners/lpc55/src/initializer.rs#L537)

Yes, we do. See spec below. Let's make sure the proper version is used.
- https://fidoalliance.org/specs/fido-v2.0-rd-20161004/fido-client-to-authenticator-protocol-v2.0-rd-20161004.html
    - search for `6.1.9.1.2 CTAPHID_INIT`

>    various RGB related changes and adaptations inside board/trussed.rs
        to be discussed if we really want a math-lib as dep to calc a LED-amplitude
        anyways RGB should be heavily tested
        added Intensities::scale_by() for easy scaling (currently misused by scaling with the amplitude...)

Yes, math lib for that looks like overkill to me as well. Why the final values were not tabularized / calculated during the build via the const func? Do you know what was wrong with the initial `sin` approximation func?
See how this was done for the Nitrokey FIDO2:
- https://github.com/Nitrokey/nitrokey-fido2-firmware/blob/master/targets/stm32l432/src/led.c#L17-L50


>    src/types.rs UsbClasses with or without serial ?
        maybe #cfg[] ? based on feature?
        serial pulled in for now...

AFAIK it got removed in the upstream, as this was mainly debug feature. Or am I wrong?
I do not see the need to leave it enabled for the release. Potentially worth to make it a feature for the fast debug purposes.

**Comment by szszszsz** [comment]
Final question: have you used the stable upstream release for the merge? Or an arbitrary commit?
And where is the CI build backlink? :|

File: runners/lpc55/.cargo/config.toml [hunk 2]
```diff
@@ -3,7 +3,7 @@ runner = "arm-none-eabi-gdb -q -x jlink.gdb"
 rustflags = [
   "-C", "linker=flip-link",
   "-C", "link-arg=-Tlink.x",
-  # "-Dwarnings",
+  "-Dwarnings",
```
[the remark this answers is not in the document]
Comment (robin-nitrokey) [reply runners/lpc55/.cargo/config.toml:6]: `-D, --deny LINT     Set lint denied` – so this should make the compilation fail if there are compiler warnings.  Makes sense for the CI, less so for development.

File: runners/lpc55/src/initializer.rs (hunk 1 above)
Comment (robin-nitrokey) [reply runners/lpc55/src/initializer.rs:86]: We had this discussion before the release and the conclusion was that we disable it by default and add an option to enable it later.

**Comment by szszszsz** [comment]
Bumping. ETA?

**Comment by szszszsz** [comment]
Is this fixing below by any chance?
- https://github.com/solokeys/fido-authenticator/issues/3

**Comment by robin-nitrokey** [comment]
> Is this fixing below by any chance?
> 
>     * [SSH: Unable to download resident keys solokeys/fido-authenticator#10](https://github.com/solokeys/fido-authenticator/issues/10) (edit)

That should already be fixed with our v1.0.1 release.  The issue we are having with resident keys is querying them, not creating them.

File: runners/lpc55/Makefile [hunk 3]
```diff
@@ -12,6 +12,14 @@ else ifeq "${PROVISIONER}" "1"
...
 endif
 
+build-release:
+	cargo build --release --features board-solo2
+	cargo objcopy --release --features board-solo2 -- -O binary app-solo2.bin
+
+.PHONY: today-as-minor-version
+today-in-days:
+	python -c "from datetime import date as d; print((d.today() - d(2020, 1, 1)).days)"
+
```
Comment (robin-nitrokey) [inline runners/lpc55/Makefile:22]: Can’t we remove this?  We don’t need it and at this position, it would be the new default target.

Comment (daringer) [reply runners/lpc55/Makefile:22]: nothing against, the `Makefile` anyways is fully rewritten inside `embedded-runner-3` due to the multiple SoCs approach. As of now something similar does not exist there...

**Comment by robin-nitrokey** [comment]
I’ve added some commits to finalize this PR – in my opinion, it is ready now.  @szszszsz @daringer Please review.

The only non-trivial change I made is the last commit.  It is necessary to maintain the current LED behavior, but it has to use a patched Trussed version because of the upstream changes.  [Here](https://github.com/Nitrokey/trussed/commit/0994607f6fe84a3fbec9b137c9e8247f1232d536) is the required patch.

**Comment by robin-nitrokey** [comment]
Also, can someone please test NFC with these changes as there seem to be new issues for solo2?

**Review by robin-nitrokey — APPROVED** [review]

File: runners/lpc55/Cargo.toml [hunk 4]
```diff
@@ -104,6 +104,9 @@ log-debug = []
 log-warn = []
 log-error = []
 
+[patch.crates-io]
+trussed = { git = "https://github.com/nitrokey/trussed", branch = "no-ui-status-reset" }
```
Comment (szszszsz) [inline runners/lpc55/Cargo.toml:108]: Do we leave side branch for now?

File: runners/lpc55/board/src/trussed.rs [hunk 5]
```diff
@@ -119,23 +108,23 @@ RGB: RgbLed,
                 ui::Status::Idle => {
                     if self.provisioner {
                         // white
-                        rgb.set(0xff_ff_ff.into());
+                        rgb.set(WHITE.into());
```
Comment (szszszsz) [inline runners/lpc55/board/src/trussed.rs:111]: nit: this feels like
> int ZERO = 0;

Better would be to have references by function:
- Color::processing
- Color::idle
- Color::idle_provisioner

etc.

File: runners/lpc55/Cargo.toml [hunk 6]
```diff
@@ -4,8 +4,8 @@
 name = "runner"
 version = "1.0.3"
 authors = ["Nicolas Stalder <n@stalder.io>", "Conor Patrick <conor@solokeys.com>"]
```
Comment (szszszsz) [inline runners/lpc55/Cargo.toml:6]: We should update authors field some day.

**Review by szszszsz — APPROVED** [review]

File: runners/lpc55/Cargo.toml (hunk 4 above)
Comment (robin-nitrokey) [reply runners/lpc55/Cargo.toml:108]: If we want to keep the FIDO2 behavior consistent, we have to.

File: runners/lpc55/board/src/trussed.rs (hunk 5 above)
Comment (robin-nitrokey) [reply runners/lpc55/board/src/trussed.rs:111]: I don’t agree – I think 0xffffff = white is non-trivial.  Anyway, we’ll change the LED behavior soon so we can address that then.

File: runners/lpc55/Cargo.toml (hunk 6 above)
Comment (robin-nitrokey) [reply runners/lpc55/Cargo.toml:6]: We can do so with the embedded runner.

**Comment by szszszsz** [comment]
@daringer Engage!

**Review by daringer — COMMENTED** [review]
+1

**Comment by robin-nitrokey** [comment]
To test:
- [ ] Check whether https://github.com/Nitrokey/nitrokey-3-firmware/issues/48 is fixed
- [ ] Re-test FIDO2 over NFC to rule out NFC issues as with latest solo2 beta firmware: https://gist.github.com/nickray/7de5d40d5d93b7a1b6d0983062c9339e

**Comment by szszszsz** [comment]
Looks good. I do not see any problems while using under the latest Windows 10. #48 is fixed.
Using Resident Keys works well, including OpenSSH.
As discussed, the following will be retested on the production hardware:
- NFC,
- User data retention.

**Pull request merged**